src/Fl_SVG_Image.cxx

```diff
@@ -90,143 +144,165 @@ float Fl_SVG_Image::svg_scaling_(int W, int H) {
-                         const unsigned char *bytes // byte buffer to write to pipe
-                         ) {
-  size_t rest_bytes = size;
-  int l;
-  size_t out_size = is_compressed ? 3 * size + 1 : size + 1;
-  char *out = (char*)malloc(out_size);
-  char *p = out;
+// Decompress gzip data in memory
+#define CHUNK_SIZE (2048)
+static int svg_inflate(uchar *src, size_t src_length, uchar *&dst, size_t &dst_length) {
+  // allocate space for decompressed data in chunks
+  typedef struct Chunk {
+    Chunk() { next = NULL; }
+    struct Chunk *next;
+    uchar data[CHUNK_SIZE];
+  } Chunk;
+  Chunk *first = NULL;
+  Chunk *chunk = NULL, *next_chunk;
+
+  z_stream stream = { };
+  int err = Z_OK;
+  const uInt max = (uInt)-1;
+
+  dst = 0;
+  dst_length = 0;
+
+  stream.next_in = (z_const Bytef *)src;
+  stream.avail_in = 0;
+  stream.zalloc = (alloc_func)0;
+  stream.zfree = (free_func)0;
+  stream.opaque = (voidpf)0;
+
+  // initialize zlib for inflating compressed data
+  err = inflateInit2(&stream, 31);
+  if (err != Z_OK) return err;
+  gz_header header;
+  err = inflateGetHeader(&stream, &header);
+  if (err != Z_OK) return err;
+
+  stream.avail_out = 0;
+  stream.avail_in = (uInt)(src_length ? src_length : -1);
+
+  // inflate into as many chunks as needed
   do {
-    if (is_compressed && p + size > out + out_size) {
-      out_size += size;
-      size_t delta = (p - out);
-      out = (char*)realloc(out, out_size + 1);
-      p = out + delta;
+    if (stream.avail_out == 0) {
+      next_chunk = new Chunk;
+      if (!first) first = next_chunk; else chunk->next = next_chunk;
+      chunk = next_chunk;
+      stream.avail_out = CHUNK_SIZE;
+      stream.next_out = chunk->data;
     }
-    if ( fdwrite >= 0 && Fl::system_driver()->write_nonblocking_fd(fdwrite, bytes, rest_bytes) ) {
-      free(out);
-      out = NULL;
-      is_compressed = false;
-      break;
+    err = inflate(&stream, Z_NO_FLUSH);
+  } while (err == Z_OK);
+
+  inflateEnd(&stream);
+
+  // copy chunk data into a new continuous data block
+  if (err == Z_STREAM_END) {
+    size_t nn = dst_length = stream.total_out;
+    dst = (uchar*)malloc(dst_length+1); // leave room for a trailing NUL
+    uchar *d = dst;
+    chunk = first;
+    while (chunk && nn>0) {
+      size_t n = nn > CHUNK_SIZE ? CHUNK_SIZE : nn;
+      memcpy(d, chunk->data, n);
+      d += n;
+      nn -= n;
+      chunk = chunk->next;
     }
+  }
 
-    l = gzread(gzf, p, (unsigned int)size);
-    if (l > 0) {
-      p += l; *p = 0;
-    }
-  } while (is_compressed && l >0);
-  gzclose(gzf);
-  if (is_compressed) out = (char*)realloc(out, (p-out)+1);
-  return out;
+  // delete all the chunks that we allocated
+  chunk = first;
+  while (chunk) {
+    next_chunk = chunk->next;
+    delete chunk;
+    chunk = next_chunk;
+  }
+
+  return err == Z_STREAM_END ? Z_OK :
+  err == Z_NEED_DICT ? Z_DATA_ERROR  :
+  err == Z_BUF_ERROR && stream.avail_out ? Z_DATA_ERROR :
+  err;
```

Comment:
missing indentation

Reply:
Copied from zlib. I will see if I can indent that in a useful way.